README.md
# SvgTracing

**Elixir Vector Graphics Converter **
An open source software to convert raster images (like jpg & png) into vector graphics (svg). It can vectorize graphics and photographs and trace the curves to output compact vector files.
This is backed by rust (vtracer) using rustler. You need to install [rust](https://www.rust-lang.org/tools/install) first.

## Installation

The package can be installed by adding `svg_tracing` to your list of dependencies in `mix.exs`:

```elixir
you need install rust 
def deps do
  [
    {:svg_tracing, "~> 0.1.4"}
  ]
end

iex(1)> SvgTracing.trace("priv/static/ikea-sofa.png", "priv/static/ikea-sofa.svg")
{:ok, {}}
```

### Examples

#### Input Image:
![input_image](/priv/static/example1.png)

#### Output SVG:
![input_image](/priv/static/example1.svg)


#### Input Image:
![input_image](/priv/static/example2.jpg)

#### Output SVG:
![input_image](/priv/static/example2.svg)


#### Input Image:
![input_image](/priv/static/example3.jpg)

#### Output SVG:
![input_image](/priv/static/example3.svg)


CHANGELOG.md

## 0.1.4 (2023-11-03)
- Update Readme

## 0.1.3 (2023-11-03)
- Add missing prive folder

## 0.1.2 (2023-11-03)
- Add missing native rust folder

## 0.1.1 (2023-11-03)
- Update Docs

## 0.1.0 (2023-09-24)
- Init project

test/svg_tracing_test.exs
defmodule SvgTracingTest do
  use ExUnit.Case
  doctest SvgTracing

  test "greets the world" do
    assert SvgTracing.hello() == :world
  end
end


mix.exs
defmodule SvgTracing.MixProject do
  use Mix.Project
  @app :svg_tracing
  @source_url "https://github.com/graydientAI/svg_tracing"

  def project do
    [
      app: @app,
      name: "svg_tracing",
      description: "Elixir Vector Graphics Converter",
      version: "0.1.4",
      elixir: "~> 1.13",
      start_permanent: Mix.env() == :prod,
      compilers: [:elixir_make | Mix.compilers()],
      make_targets: ["all"],
      make_clean: ["clean"],
      deps: deps(),
      package: package(),
      releases: [{@app, release()}]
    ]
  end

  # Run "mix help compile.app" to learn about applications.
  def application do
    [
      extra_applications: [:logger],
      mod: {SvgTracingScript, []}
    ]
  end

  # Run "mix help deps" to learn about dependencies.
  defp deps do
    [
      {:rustler, "~> 0.29.1"},
      {:elixir_make, "~> 0.6.1"},
      {:bakeware, "~> 0.2.4", runtime: false},
      # {:bakeware, path: "../..", runtime: false},
      {:ex_doc, ">= 0.0.0", only: :dev, runtime: false}
    ]
  end

  defp release do
    [
      overwrite: true,
      cookie: "#{@app}_cookie",
      quiet: true,
      steps: [:assemble, &Bakeware.assemble/1],
      strip_beams: Mix.env() == :prod
    ]
  end

  defp package do
    [
      maintainers: ["Hassan Raza"],
      licenses: ["MIT"],
      links: %{GitHub: @source_url},
      canonical: "https://hexdocs.pm/svg_tracing",
      files: ~w(lib native priv .formatter.exs mix.exs README.md LICENSE CHANGELOG.md)
    ]
  end
end


native/svgtracing_tracer/src/lib.rs
use visioncortex::PathSimplifyMode;
use vtracer::{Config, convert_image_to_svg, ColorMode, Hierarchical};
use std::path::PathBuf;

#[rustler::nif(schedule = "DirtyCpu")]
fn convert_to_svg(
  image_path: &str,
  out_path: &str,
  colormode: Option<&str>,       // "color" or "binary"
  hierarchical: Option<&str>,    // "stacked" or "cutout"
  mode: Option<&str>,            // "polygon", "spline", "none"
  filter_speckle: Option<usize>, // default: 4
  color_precision: Option<i32>,  // default: 6
  layer_difference: Option<i32>, // default: 16
  corner_threshold: Option<i32>, // default: 60
  length_threshold: Option<f64>, // in [3.5, 10] default: 4.0
  max_iterations: Option<usize>, // default: 10
  splice_threshold: Option<i32>, // default: 45
  path_precision: Option<u32>,   // default: 8
) -> Result<(), ()> {
  let input_path = PathBuf::from(image_path);
  let output_path = PathBuf::from(out_path);

  // TODO: enforce color mode with an enum so that we only
  // accept the strings 'color' or 'binary'
  let color_mode = match colormode.unwrap_or("color") {
      "color" => ColorMode::Color,
      "binary" => ColorMode::Binary,
      _ => ColorMode::Color,
  };

  let hierarchical = match hierarchical.unwrap_or("stacked") {
      "stacked" => Hierarchical::Stacked,
      "cutout" => Hierarchical::Cutout,
      _ => Hierarchical::Stacked,
  };

  let mode = match mode.unwrap_or("spline") {
      "spline" => PathSimplifyMode::Spline,
      "polygon" => PathSimplifyMode::Polygon,
      "none" => PathSimplifyMode::None,
      _ => PathSimplifyMode::Spline,
  };

  let filter_speckle = filter_speckle.unwrap_or(4);
  let color_precision = color_precision.unwrap_or(6);
  let layer_difference = layer_difference.unwrap_or(16);
  let corner_threshold = corner_threshold.unwrap_or(60);
  let length_threshold = length_threshold.unwrap_or(4.0);
  let splice_threshold = splice_threshold.unwrap_or(45);
  let max_iterations = max_iterations.unwrap_or(10);

  let config = Config {
      input_path,
      output_path,
      color_mode,
      hierarchical,
      filter_speckle,
      color_precision,
      layer_difference,
      mode,
      corner_threshold,
      length_threshold,
      max_iterations,
      splice_threshold,
      path_precision,
      ..Default::default()
  };

  convert_image_to_svg(config).unwrap();
  Ok(())
}

rustler::init!("Elixir.SvgTracing.Tracer", [convert_to_svg]);


lib/svg_tracing.ex
defmodule SvgTracing do
  alias Toml.Lexer.String
  alias SvgTracing.Tracer

  @moduledoc ~S"""
  An open source software to convert raster images (like jpg & png) into vector graphics (svg).
  It can vectorize graphics and photographs and trace the curves to output compact vector files.
  This is backed by rust (vtracer) using rustler.

  iex> SvgTracing.trace("priv/static/ikea-sofa.png", "priv/static/ikea-sofa.svg")
  {:ok, {}}
  """
  defstruct input_path: "",
            output_path: "",
            colormode: "color",
            hierarchical: "stacked",
            mode: "spline",
            filter_speckle: 4,
            color_precision: 6,
            layer_difference: 16,
            corner_threshold: 60,
            length_threshold: 4.0,
            max_iterations: 10,
            splice_threshold: 45,
            path_precision: 8

  @doc """
  Convert images to vector graphics, this will need two paths, input image path
  and second output image path to save svg.
  
  iex(1)> SvgTracing.trace("priv/static/ikea-sofa.png", "priv/static/ikea-sofa.svg")
  {:ok, {}}
  """
  @spec trace(String.t(), String.t()) :: {:ok, {}} | {:error, String.t()}
  def trace(input_path, output_path) do
    default_values = %SvgTracing{}

    Tracer.convert_to_svg(
      input_path,
      output_path,
      default_values.colormode,
      default_values.hierarchical,
      default_values.mode,
      default_values.filter_speckle,
      default_values.color_precision,
      default_values.layer_difference,
      default_values.corner_threshold,
      default_values.length_threshold,
      default_values.max_iterations,
      default_values.splice_threshold,
      default_values.path_precision
    )
  end
end
